### PR TITLE
Fixing incorrect line reporting in Scintilla controls when line wrapping is on.

### LINKED ISSUE #9424: NVDA's support for Scintilla controls. is broken in alpha snapshots, if "line wrap" is enabled in the text editor
### Steps to reproduce:
1. Install a text editor like Notepad++.
2. In it, open a text file containing a few lines of text.
3. Enable the text editor's "line wrap" feature.
4. Start reading line by line with the up/down arrow keys.
5. Enable NVDA's mouse tracking and use the mouse to hover over and read the separate lines of text.

### Actual behavior:
With the up/down arrow keys, NVDA reads another line of text and not the one on which the cursor is positioned. This behavior affects only reading line by line - reading by word or character seams to be working as expected. The situation when reading with the mouse is similar - NVDA does not read the line which the mouse is pointing, but another one. And this occurs only if the "line wrap" feature of the text editor is enabled, so long lines of text are wrapped on multiple lines if necessary.

### Expected behavior:
NVDA should read the line on which the cursor is positioned or the one which the mouse is pointing, even if the "line wrap" feature of the text editor is enabled.

### System configuration
- Operating system: Windows 10 Pro version 1809 (build 17763.379), 64-bit, in Bulgarian with all locale settings set to "Bulgarian".
- NVDA version: alpha-16976,7fc6c71f.
- Processor: Intel Core i5-4670 at 3.40GHz.
- RAM Memory: 8.00GB.
- Graphics: MSI GeForce GTX-1050TI Gaming X 4G, 4096MB dedicated memory + 4065MB shared system memory, desktop resolution set to 1920x1080.
- Sound Card: Realtek ALC887 at Intel Lynx Point PCH - High Definition Audio Controller.

#### Does the issue still occur after restarting your PC?
Yes, it does.

#### Have you tried any other versions of NVDA?
No, because this started occurring only with the latest "alpha" snapshots of NVDA. I guess this is probably caused by #9233 (Fixing incorrect text reporting when hovering mouse over Scintilla controls), though I might be wrong.

### PR DESCRIPTION
**Pull request by DataTriny**
### Link to issue number:

Fixes #9424.

### Summary of the issue:

When navigating inside a Scintilla control with line wrapping enabled, NVDA fails to report the correct line (either when navigating with the keyboard or the mouse).

### Description of how this pull request fixes the issue:

In `_getLineOffsets`:

- Firstly `self.obj.location` were not properly deconstructed (top and left were swapped),

- `_getOffsetFromPoint` were called with `x=0` to retrieve the start point of the line, ignoring the real x coordinate of the Scintilla control,

- finally, `_getOffsetFromPoint` were called with `x=width` to retrieve the endpoint of the line, again ignoring the margin on the left of the control.

### Testing performed:

Navigated through a document in Notepad++ with line wrapping enabled. Made sure that it worked with multiple DPI factors, on short lines as well as wrapped ones.

### Known issues with pull request:

None

### Change log entry:

None

### CODE REVIEW

File: source/NVDAObjects/window/scintilla.py
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]: note that you can make use of locationHelper here, so you do not have to extract the tuple.

```suggestion
			location = self.obj.location
```

File: source/NVDAObjects/window/scintilla.py
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]:
```suggestion
			start = self._getOffsetFromPoint(location.left, location.top + y)
```

File: source/NVDAObjects/window/scintilla.py
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]:
```suggestion
			end=self._getOffsetFromPoint(location.right, location.top + y)
```

**Review by LeonarddeR — APPROVED** [review]
Thanks for testing this thoroughly, also with several DPI levels. I will push a try build for people to test.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Actually, after reading the code again, there is something that seems odd to me.

1. We're fetching the location of the control, which is in screen coordinates.
2. Then, when calling self._getOffsetFromPoint, you are adding the result of SCI_POINTYFROMPOSITION to the top of the window. However, according to the Scintilla docs:

    > These messages return the x and y display pixel location of text at position pos in the document.

3. I also noticed that _getOffsetFromPoint throws client coordinates at scintilla, wwhereas _getPointFromOffset doesn't do anything with client to screen conversion. Could you look into this while at it?

File: source/NVDAObjects/window/scintilla.py
Comment (DataTriny) [reply source/NVDAObjects/window/scintilla.py]: This way might be less error prompt and is a bit clearer.

**Comment by DataTriny** [comment]
@leonardder Regarding your second point: it also sounded strange to me at first, but I tried several combinations and only this one was giving the right results. I claim that it's pixel perfect!

Finally, `_getPointFromOffset` already returns screen coordinates according to my testing (even if I didn't thought). But I can't figure out where this function gets called so I tested it manually.

**Comment by leonardder** [comment]
_getPointFromOffset gets called when you use the move mouse to navigator object command. My testing seemed to reveal some issues with this..

**Comment by DataTriny** [comment]
Hi again @leonardder,

So... `SCI_POINTXFROMPOSITION` and `SCI_POINTYFROMPOSITION` unfortunately don't give screen coordinates. I had to add the left and top margin of the control.

Anything else?

File: source/NVDAObjects/window/scintilla.py
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]: You're basically converting client to screen coordinates here, doing it manually. Is there a specific reason why you aren't using clientToScreen here? Does it fail?

Comment (DataTriny) [reply source/NVDAObjects/window/scintilla.py]: Hi Leonard,

You are right! I just forgot about it. This way makes it more explicit.

File: source/NVDAObjects/window/scintilla.py
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]: You now can retrieve y with _getPointFromOffset. That gives you an x you won't use, but it at least ensures that you're following an uniform codepath to get screen coordinates from scintilla offsets.

File: source/NVDAObjects/window/scintilla.py
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]: If you use _getPointFromOffset above, this will be:

```suggestion
			start = self._getOffsetFromPoint(location.left, y)
```

File: source/NVDAObjects/window/scintilla.py
Comment (leonardder) [inline source/NVDAObjects/window/scintilla.py]: When using _getPointFromOffset above, this becomes:

```suggestion
			end=self._getOffsetFromPoint(location.right, y)
```

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Almost there

File: source/NVDAObjects/window/scintilla.py
Comment (DataTriny) [reply source/NVDAObjects/window/scintilla.py]: Hi Leonard,

Yes it makes sense. I initially didn't write it like this because I thought a useless call to `watchdog.cancellableSendMessage` to get the X coordinate wouldn't be accepted. But I don't really know how expensive is that type of call.

**Comment by leonardder** [comment]
While the current approach involves more system calls, I think it is much less error prone. I'll have another look at your code later this week and then it's ready for Mick to get looked at.

**Review by LeonarddeR — APPROVED** [review]

**Comment by dingpengyu** [comment]
hi
What is the current status of this PR? I saw that he had been approved.

**Comment by leonardder** [comment]
I have approved it, but approval from a NV Access member is required before this can be merged.

**Comment by ABuffEr** [comment]
Hi,
I have no role here, but I kindly pray @michaelDCurran to review and merge this pr, or revert the changes that had broken line wrapping support. Using Notepad++ each day for multiple purposes, I must use a stable, without being able to test new features in alpha, very annoying situation.

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**
